src/main/ApplicationImpl.cpp

```diff
@@ -285,6 +285,10 @@ ApplicationImpl::~ApplicationImpl()
     if (mWorkScheduler)
     {
         mWorkScheduler->shutdown();
+        while (mWorkScheduler->getState() != BasicWork::State::WORK_ABORTED)
```

Comment:
this makes me think:

WorkScheduler never gets into WORK_SUCCESS or WORK_FAILURE states

It also never uses most of Work/BasicWork APIs. So it is really a Work? Maybe it would be better for it to have an internal never-ending Work item and make all scheduled Works children of that one?

Reply:
What I mean is that currently:

WorkScheduler is Work is BasicWork

which means that WorkScheduler API looks like this:

```
BasicWork::getName
BasicWork::getStatus
BasicWork::getState
BasicWork::isDone
BasicWork::crankWork
BasicWork::startWork
BasicWork::shutdown
BasicWork::isAborting
Work::allChildrenSuccessful()
Work::allChildrenDone()
Work::anyChildRaiseFailure()
Work::anyChildRunning()
Work:: hasChildren()
WorkScheduler::executeWork
WorkScheduler::scheduleWork
```

And the used part is:
```
BasicWork::getState (but we only check for WORK_ABORTED and WORK_WAITING in tests)
Work::allChildrenDone (tests)
Work::allChildrenSuccessful (tests)
WorkScheduler::shutdown
WorkScheduler::executeWork
WorkScheduler::scheduleWork
```

I also think that calling `startWork` or `crankWork` on this class would not be a good idea.

This is why I propose to hide all that Work/BasicWork API from users of WorkScheduler either by making its Work inheritance private, or making it not inherit from Work and just use Work instance as a member internally and expose specially-tailored API just for WorkScheduler.

Reply:
I'm just not quite convinced by the argument that WorkScheduler is not a Work. In my opinion, this is an `is-a` relationship rather than `has-a`. Besides 5 main methods in the API, the rest are just const utility methods, which could be used by the scheduler in the future. I also think that calling `startWork` and `crankWork` is just as bad for classes like BasicWork/Work (that's why those methods are equipped with strict state transition mechanisms), so in general, the user is forced to use them correctly.